docs/design-documents/platform/crash-reporting/crash_reporting.md

```diff
@@ -0,0 +1,276 @@
+
+    Do a system reset //using system_reset() function
+}
+
+```
+The Crash-Report RAM region should also be used for tracking other pieces of information such as the CRC value and the
+auto-reboot count. 
+
+### Mechanism to retrieve and reset the error data after reboot
+
+MbedOS error handling system should implement necessary APIs for application to retrieve and reset the error and/or fault context of the previous fatal error after auto-reboot.
+The below APIs should be implemented.
+
+The below API can be called by application to retrieve the error context captured in the Crash-Report RAM. The error context is copied into the location pointed by *error_info*. Note that the caller should allocate the memory for this location.
+The function should return MBED_ERROR_NOT_FOUND if there is no error context currently stored.
+```C
+//Retrieve the reboot error context
+mbed_error_status_t mbed_get_reboot_error_info(mbed_error_ctx *error_info)
+```
+
+The below API can be called by application to retrieve the fault context captured in the Crash-Report RAM. The error context is copied into the location pointed by *fault_context*. Note that the caller should allocate the memory for this location. Note that the fault context is valid only if the previous reboot was caused by an exception. Whether the previous reboot was caused by an exception can be determined from the error code stored in error context information retrieved using mbed_get_reboot_error_info() API above.
+The function should return MBED_ERROR_NOT_FOUND if there is no fault context currently stored.
+```C
+//Call this function to retrieve the last reboot fault context
+mbed_error_status_t mbed_get_reboot_fault_context (mbed_fault_context_t *fault_context);
+```
+
+The below API can be called by application to reset the error context captured in the Crash-Report RAM.
+The function should MBED_ERROR_NOT_FOUND if there is no error context currently stored.
+```C
+//Reset the reboot error context
+mbed_error_status_t mbed_reset_reboot_error_info()
+```
+
+### Mechanism to report the error data after reboot
+
+MbedOS initialization sequence should check if the reboot is caused by a fatal error and should report the same to the application using callback mechanism. The system may also print this to terminal if enabled. 
+
+MbedOS initialization sequence should be modified as shown in below diagram to report the crash report and invoke the callback.
+
+![Error report on reboot](./diagrams/boot-error-report.jpg)
+
+Below should be the signature of the callback for reporting the error information.
+
+The error handing system in MbedOS will call this callback function if it detects that the current reboot has been caused by a fatal error. This function will be defined with MBED_WEAK attribute by default and applications wanting to process the error report should override this function in application implementation.
+```CS
+void mbed_error_reboot_callback(mbed_error_ctx *error_context);
```

Comment:
Slightly lost on the twin callback + read APIs here. Not seeing why this callback is necessary. Can't the application just call `mbed_get_reboot_fault_context` if they want to find out if it was a reboot-due-to-error? One example suggests that, but another example shows it taking a copy in the callback, as if it's going to be lost.

Reply:
The callback is necessary for few reasons. It provides opportunity for the application to clear any error situation based on error codes before the main starts. Its also required in the case where a reboot-max is configured where in the system will halt at the maximum reboot count, but even in this case callback will be invoked, so the app can know the cause of error. Also, if the application design doesn't require the callback, they don't have to override that, so its flexible. Also making calls to retrieve the error context and fault context could be expensive as well as they take time during main() initialization and also the caller needs to allocate memory. So having this callback to set a flag(or something like that) can bring some optimization as well.